---
title: "Script 1_GTex_DESeq2_DEAnalysis"
author: "Nickie Safarian"
date: "11/29/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This R Markdown document contains explains DE analysis for GTex_phs000424 
dataset using DESeq2 package.

```{r, include=FALSE}


library(tidyverse)
library(cowplot)
library(knitr)
library(DESeq2)

```

## Imort the Metadata

The following chunks read in an older version (V8) of expression counts and metadata:

```{r, echo=FALSE, include=FALSE}

# 1) get sample.metadata from gtex website
gtex_samples_meta = read_delim(url("https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"), delim = '\t')


# 2) get phenotype.metadata (this file only contains Age and sex vars)
# gtex_pheno_meta = read_delim(url('https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt'), delim = '\t')


gtex_subjects_meta = read_delim('/external/rprshnas01/kcni/nsafarian/GTex_Exp_Analysis/Data/urlDownload/phs000424.v8.pht002742.v8.p2.c1.GTEx_Subject_Phenotypes.GRU.txt.gz', delim = '\t', skip = 10)


```


```{r}

# Check the file in this url to learn what the metadata columns headings code for:

url= 'https://view.officeapps.live.com/op/view.aspx?src=https%3A%2F%2Fstorage.googleapis.com%2Fgtex_analysis_v8%2Fannotations%2FGTEx_Analysis_v8_Annotations_SampleAttributesDD.xlsx&wdOrigin=BROWSELINK'

# RIN= SMRIN
# PMI (in min)= SMTSISCH
# SMTSD = Tissue.Type 

#url for phenotypes
'https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/variable.cgi?study_id=phs000424.v7.p2&phv=169064&phd=3910&pha=&pht=2742&phvf=&phdf=&phaf=&phtf=&dssp=1&consent=&temp=1'
```

## Preprocess the Metadata

```{r}

# get samples from specific brain regions (i.e., cortex)
use_cortex_regions = c("Brain - Frontal Cortex (BA9)")

# filter sample.metadata to only include the region of interest
use_samples = gtex_samples_meta %>% filter(SMTSD %in% use_cortex_regions) %>% pull(SAMPID) #425 samples are from BA9

# subset the metadata for only BA9 samples
brain_gtex_metadata = gtex_samples_meta %>% filter(SAMPID %in% use_samples) 

```

In order to combine the samples.metadata and subject.metadata, we need to match the ID columns between the two data sets. To do so

```{r, echo=FALSE, include=FALSE}

#  add a column called SUBJID to metadata that enables linking with subject metadata
brain_gtex_metadata = brain_gtex_metadata %>% 
  separate(SAMPID, into = c('first', 'second'), remove = F, sep = '-') %>% 
  unite(col = "SUBJID", first:second, sep = '-', remove = F) 


# add age and sex column to the previous metadata
METADATA <- merge(gtex_subjects_meta, brain_gtex_metadata, 
                  by.x='SUBJID', by.y = 'SUBJID') #425 IDs

rownames(METADATA) <- METADATA$SAMPID

```

## Import the counts matrix

```{r, include=FALSE}

# read in full RNAseq reads - careful, this file is  huge (14 GB)
reads_data_file_name = '/external/rprshnas01/netdata_kcni/stlab/Public/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct'

gtex_reads = read_tsv(paste0(reads_data_file_name), skip=2)


kable(gtex_reads[1:3, 1:4]) # 56200 * 17384 

```


```{r}

# remove the Gene.Symbol column from the count matrix
gene.symbol <- gtex_reads[,c(1, 2)]

# subset counts/reads for the select samples
get_sample_col_names = intersect(c('Name', use_samples), colnames(gtex_reads))

brain_gtex_reads = gtex_reads[ ,get_sample_col_names] # 211 cols

# Set Ensembl IDs column as rownames
brain_gtex_reads = brain_gtex_reads %>% column_to_rownames(var= 'Name')

```


#----------------------------------------------------------------------------

## DESEq2 steps


```{r, include=FALSE}

# match Ids between metadata and counts
coldata = METADATA%>% subset(SAMPID %in% colnames(brain_gtex_reads)) #209

# Very important:
cts <- round(brain_gtex_reads) %>%  # convert values to integers
  as.data.frame() %>%
  dplyr::filter(rowSums(.) > 0) # remove genes with zero total 
                                 # counts across samples

```

### *Check coldata & Cts IDs are matching and in the same order*

```{r}

# Ensure coldata and cts are in the same order
all(rownames(coldata) == colnames(cts))
# Ensure the same samples across cts and coldata
all(rownames(coldata) %in% colnames(cts))

```

### *check for correlations*

```{r}

num.vars <- coldata %>% select (c("AGE", "SMRIN","MHBRNPH", "SMTSISCH",  "TRISCHD")) # Age, RIN, pH, PMI1, PMI2

CovarCor = cor(num.vars, method = "pearson", use = "complete.obs")
round(CovarCor, 2)

```

```{r}

cormat <- cor(num.vars)
cormat %>% as.data.frame %>% mutate(var=rownames(.)) %>%
  pivot_longer(!var, values_to = "value") %>%
  ggplot(aes(x=name,y=var,fill=abs(value),label=round(value,2))) +
  geom_tile() + geom_label() + xlab("") + ylab("") +
  ggtitle("Correlation matrix of our predictors") +
  labs(fill="Correlation\n(absolute):")


```


### *Prepare covariates in the coldata accordingly*

```{r}

coldata$AGE <- scale(coldata$AGE)
coldata$SMRIN <- scale(coldata$SMRIN) # this is RIN
coldata$TRISCHD <- scale(coldata$TRISCHD) # this is probably PMI

# first rename levels for SEX (there is level 2 in both sex and race factors, so one has to change before making DESeq2 object)
coldata$SEX[coldata$SEX=="1"] <- "male"
coldata$SEX[coldata$SEX=="2"] <- "female"

coldata$RACE[coldata$RACE=="2"] <- "2A"
coldata$RACE[coldata$RACE=="3"] <- "3B"


coldata$RACE = factor(coldata$RACE, c("2A", "3B"))
coldata$SEX <- factor(coldata$SEX , c("male","female"))


```


```{r}

dds <- DESeqDataSetFromMatrix(countData = cts, 
                              colData = coldata, 
                              design = ~ RACE+SEX+TRISCHD+SMRIN+AGE)
```

```{r}

# Perform pre-filtering to remove genes that are lowly expressed 
# (at least 10 counts)
dds <- estimateSizeFactors(dds)
ids <- rowSums(counts(dds, normalized=TRUE) >= 10 ) >= 3 
dds <- dds[ids, ]

```

```{r}

dds$RACE <- relevel(dds$RACE, '3B')
dds$SEX <- relevel(dds$SEX, 'male')

```

```{r}

dds <- DESeq(dds)

```

```{r}

resultsNames(dds) 
```

```{r}

GTex.Res.Age <- results(dds, name="AGE")
summary(GTex.Res.Age)

```




```{r}

sigs <- na.omit(GTex.Res.Age) 

sigs <- sigs[sigs$padj <0.05, ]  # you may not do this step if you will check for Age_NPs expression at the end.


sigs.df <- as.data.frame(sigs) %>%
  rownames_to_column(var= 'EnsemblID') #7069 genes

```

### *Add gene names*

```{r}

sigs.df$EnsemblID <- sapply(strsplit(sigs.df$EnsemblID, split="\\." ), "[", 1 )


#The following chunk of code uses the ENSEMBL mart, querying with the ENSEMBL gene id and requesting the Entrez gene id and HGNC gene symbol.

library( "biomaRt" )
ensembl <- useMart( "ensembl", dataset = "hsapiens_gene_ensembl" )
genemap <- getBM( attributes = 
                    c("ensembl_gene_id", "hgnc_symbol"),
                  filters = "ensembl_gene_id", 
                  values = sigs.df$EnsemblID, 
                  mart = ensembl)


idx <- match(sigs.df$EnsemblID, genemap$ensembl_gene_id )
sigs.df$entrez <- genemap$external_gene_name[ idx ]
sigs.df$hgnc_symbol <- genemap$hgnc_symbol[ idx ]


```


```{r}
# save
write.csv(sigs.df, "gtex_v8_AGE_related_sigs_DEGs.csv")


```


```{r}
#save 
write.csv(gtex_samples_meta, "gtex_samples_metadata.csv")
write.csv(gtex_subjects_meta, "gtex_Subjects_metadata.csv")
write.csv(METADATA, "getx_MetadataFull_ONLY.BA9.Samples.csv")

saveRDS(gtex_reads, "gtexV8_RNAseq.Reads.Rds")
saveRDS(brain_gtex_reads, "gtexV8_RNAseq.Reads_ONLY.BA9.Samples.Rds")

saveRDS(dds, "GTeX_V8_DESeq2object.dds_209SamplesBA9.Rds")

```

## Check for NPs gene expression levels

```{r}

# Vanessa's Reported NPs list
Age_NPs <- read.csv(file=file.path("/external/rprshnas01/kcni/nsafarian/CMC_Exp_Analysis/Neuropeptides/Age_NPs_withEnsemblID.csv")) # 81 NPs


matchingNPs <- sigs.df %>% subset(EnsemblID %in% Age_NPs$ensembl_gene_id)

#save
write.csv(matchingNPs, "List.of.Age-related.NPs.in.GTeX.csv")

```

# Note:
From the Etienne's Age_NPs list 64 matched the GTex_Age_DEGs result. Though, of
these 64 NPs, only 25 are significant tags (i.e., FDR<0.05). 

For SST gene --> log2FoldChange= -0.36045913, padj=2.781753e-04






